compiler/resolution/functionResolution.cpp

```diff
@@ -4985,7 +4985,13 @@ static void resolveNew(CallExpr* call) {
             if (isBlockStmt(call->parentExpr) == true) {
               call->insertBefore(def);
             } else {
-              call->parentExpr->insertBefore(def);
+              Expr* parent = call->parentExpr;
```

Comment:
I believe both "in a block" and "on a record" need to be true for the transformation I'm inserting to be necessary - if the inserted init call started as a top-level statement, then we weren't expecting it to return anything in the first place.

It might be worthwhile to split the handling of non-generic classes and non-generic records out from each other, to allow the control flow in this area to be more straight-forward.  I'm happy to do that in this PR if you agree